Refactor of editors

pmk65 opened the pull request:
Refactored `src/editors/string.js` with Ace, SimpleMDE and SCEditor code removed and place into separate edtors. (See #400)
Each editor is pretty much identical and now respects the enable/disable & readonly options. 
Each editor can now be configured globally with schema option override,

~~I have kept the original code in `src/editors/string.js` for now (commented out) so don't merge it yet.~~

I also modified the `setValue()` function in `src/editors/string.js` so it returns an object in case value is set. That way the `setValue()` function in each extended editor is reduced to a couple lines.

pmk65 commented:
Test schema:
````json
{
  "$id": "https://example.com/arrays.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "jodit": {
      "type": "string",
      "format": "wysiwyg",
      "options": {
        "hidden": true,
        "jodit": {
          "height": 400,
          "theme": "dark"
        }
      }
    },
    "simplemde": {
      "type": "string",
      "format": "markdown",
      "options": {
        "simplemde": {
          "hideIcons": ["guide", "heading"]
        }
      }
    },
    "ace": {
      "type": "string",
      "format": "javascript",
      "options": {
        "ace": {
          "theme": "ace/theme/monokai"
        }
      }
    },
    "sceditor": {
      "type": "string",
      "format": "xhtml",
      "options": {
        "sceditor": {
          "toolbar": "bold,italic,underline|source"
        }
      }
    }
  }
}
````

schmunk42 commented on src/editors/string.js:
```diff
@@ -31,7 +31,7 @@ JSONEditor.defaults.editors.string = JSONEditor.AbstractEditor.extend({
     this.input.value = sanitized;
     
-    // If using SCEditor, update the WYSIWYG
+/*    // If using SCEditor, update the WYSIWYG
```
Please remove the comments ;)

pmk65 replied:
>
> I have kept the original code in src/editors/string.js for now (commented out) so don't merge it yet. 😉

I will removed it, if everyone agrees to the changes. As it changes how you configure the "plugins". (Now uses `JSONEditor.defaults.options` and schema options instead of `JSONEditor.plugins`)

(I'm still tweaking the code, and I keep the comments as reference until I'm done.)

schmunk42 replied:
> I will removed it, if everyone agrees to the changes.
> ...
> (I'm still tweaking the code, and I keep the comments as reference until I'm done.)

You can't really understand the changes in the diff, if you do not see which lines are removed.

pmk65 replied:
Good point. I have committed a version without the comments.

pmk65 commented on src/editors/string.js:
```diff
@@ -369,15 +372,15 @@ JSONEditor.defaults.editors.string = JSONEditor.AbstractEditor.extend({
         this.ace_editor = window.ace.edit(this.ace_container);
-        
+
         var aceOptions = this.schema.options && this.schema.options.ace;
```
This takes options from top level AND ace sublevel, which is a mess.

pmk65 commented:
If the changes in how the plugins are configured is accepted, I'll update documentation and examples.

I don't think that many users have bothered with custom configurations, specially if they didn't have any effect like the ones for the ACE editor.

schmunk42 commented:
Are there breaking changes for configs? Looked to me that "just" the code is now cleaner separated.

pmk65 commented:
> 
> 
> Are there breaking changes for configs? Looked to me that "just" the code is now cleaner separated.

Yes it is breaking changes. The current global config methods for ACE, SCEditor and SimpleMDE is done using `JSONEditor.plugins` (like `JSONEditor.plugins.SimpleMDE`) Some also uses options from schema and some like ACE is just fucked up. (returns a Boolean instead of the options)
````javascript
var aceOptions = this.schema.options && this.schema.options.ace;
        if (aceOptions) {
          this.ace_editor.setOptions(aceOptions);
        }
````
In the new editors (This PR), the global config is done using `JSONEditor.defaults.options` (Like `JSONEditor.defaults.options.simplemde`), which in turn can be overridden by setting schema options.

pmk65 commented:
I'm pretty sure I can refactor `src/editors/select.js` and it's plugins in a similar way as Select2 code is embedded directly in `src/editors/select.js` 

Currently `src/editors/selectize.js` and `src/editors/choices.js` entend `JSONEditor.AbstractEditor`, but 90%+ of the code is identical to the code in  `src/editors/select.js`. So it would reduce duplicate code if they extend `JSONEditor.defaults.editors.select` instead.

schmunk42 commented:
Sounds great, just one thing: if we have breaking changes, we have to create a 2.0 release.
I'd just go ahead on master and maintain 1.x on a separate branch if needed.

pmk65 commented:
Refactored src/editors/select.js too. So now Select2 is in a separate editor and is triggered by setting "format": "select2". This means you can mix normal select boxes with select2 without any problems.

Test schema:
````json
{
  "$id": "https://example.com/arrays.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "select": {
      "type": "string",
      "enum": [
        "Red",
        "Green",
        "Blue"
      ]
    },
    "select2": {
      "type": "string",
      "format": "select2",
      "enum": [
        "Red",
        "Green",
        "Blue"
      ]
    }
  }
}
````

schmunk42 commented:
@germanbisurgi Since these are BC breaking changes, could you create a test which passes on 1.4.0-beta1, but fails here.

pmk65 commented:
> 
> 
> @germanbisurgi Since these are BC breaking changes, could you create a test which passes on 1.4.0-beta1, but fails here.

Here's a list of the differences.
* Select2, Selectize and Choices are no longer applied globally to all select elements, but have to be enabled by setting "format" to the appropriate editor.
* There's no longer any `JSONEditor.plugins` options to enable Selectize or Select2.
* `JSONEditor.plugins` is obsolete and is replaced with options in `JSONEditor.defaults.options` 
* All editors can be configured globally using `JSONEditor.defaults.options`  and/or separately by schema options.
* All editors respects the enable/disable states.

pmk65 commented:
@germanbisurgi multiselect and selectize now fails the Travis tests. 
I have added validation functionality to multiselect (which was missing) and rewritten Selectize so that it uses a correct schema with enum. (To ensure fallback)
Could you check if those changes are the cause of the Travis test errors?

pmk65 commented:
My testing schema for multiselect/select2 and selectize.
````json
{
  "type": "object",
  "title": "Text",
  "required": [
    "text",
    "fontSize",
    "color",
    "font",
    "tags",
    "tags2",
    "possibleFonts"
  ],
  "properties": {
    "text": {
      "type": "string"
    },
    "fontSize": {
      "type": "integer",
      "format": "selectize",
      "enum": [
        10,
        11,
        12,
        14,
        16,
        18,
        20,
        22,
        24,
        36,
        48,
        60,
        72,
        100
      ],
      "default": 24,
      "options": {
        "select2_options": {
          "width": "off"
        }
      }
    },
    "color": {
      "type": "string",
      "enum": [
        "black",
        "red",
        "green",
        "blue",
        "yellow",
        "orange",
        "purple",
        "brown",
        "white",
        "cyan",
        "maagenta"
      ]
    },
    "font": {
      "type": "string",
      "enumSource": "possible_fonts",
      "watch": {
        "possible_fonts": "root.possibleFonts"
      }
    },
    "tags": {
      "type": "array",
      "uniqueItems": true,
      "minItems": 2,
      "options": {
        "infoText": "Infotext 1"
      },
      "items": {
        "type": "string",
        "enum": [
          "bold",
          "italic",
          "smallcaps"
        ]
      }
    },
    "tags2": {
      "type": "array",
      "format": "selectize",
      "uniqueItems": true,
      "minItems": 3,
      "options": {
        "infoText": "Infotext 1",
        "selectize": {
          "create": false
        }
      },
      "items": {
        "type": "string",
      "options": {
        "enum_titles": ["Bold", "Italics", "Small Caps"]
      },
        "enum": [
          "bold",
          "italic",
          "smallcaps"
        ]
      }
    },
    "possibleFonts": {
      "type": "array",
      "uniqueItems": true,
      "format": "selectize",
      "options": {
        "infoText": "Infotext 2"
      },
      "items": {
        "type": "string",
        "enum": [
          "Arial",
          "Times",
          "Courier",
          "Helvetica",
          "Comic Sans",
          "Open Sans"
        ]
      },
      "default": [
        "Arial",
        "Times",
        "Helvetica",
        "Comic Sans"
      ]
    }
  }
}
````

pmk65 commented:
Not sure why the `Selectize` tests keep failing. Maybe it's because I rewrote how it works?

BTW: I'm trying to come up with a good way to describe each editor and how to trigger and configure it.
If anyone has any ideas to improve it, please let me know as I suck in writing documentation 😀 
Preliminary version: https://github.com/json-editor/json-editor/blob/24fbf13997e0db2de76aaf948b2424038bbd1453/README_ADDON.md

pmk65 commented:
@schmunk42 @germanbisurgi Everything should be working now, so please go ahead and test/review the changes. 
All follow the same pattern when it comes to configuration options. 
Special editors now has to be triggered using "format" in schema. (incl. selectize and select2)
Select2 & Selectize now works with enum lists (doesn't render blank field) and supports `user tagging` (new values can be added to enum list by user if tagging option is enabled)

schmunk42 commented:
This is nice work!

@germanbisurgi Please take a look at it, so we can proceed merging it for a 2.x release.

pmk65 commented:
@germanbisurgi @schmunk42 I have made several Playground test/examples for the new and refactored editors. 
Let me know if you want me to upgrade the Playground page with the 2.x tests/examples if that helps reviewing the changes?

pmk65 commented:
@schmunk42 We need more active reviewers. How about inviting some more people into the project?
@sirockin and @itsoli seems to have a good JavaScript knowledge and insight in the codebase.

schmunk42 commented:
> @schmunk42 We need more active reviewers. How about inviting some more people into the project?
> @sirockin and @itsoli seems to have a good JavaScript knowledge and insight in the codebase.

I'd be fine with it. @sirockin and @itsoli Would that be OK for you?

schmunk42 commented:
I changed the base branch to `develop/2.x`, so we could merge this IMHO.

pmk65 commented:
There is the issue with the tests failing for the select editors. But I think it's because I rewrote the functionality of those quite a bit and fixed several bugs in the process.
So we need to test if the select based editors work as expected. (I'll update the Playground page with some new test/examples I made.)

And then I have "destroyed" @itsoli Choices editor when refactoring it. 😄 
I refactored it using same approach as the other select based editors. But then noticed that it required a "text" input field for "tags" . So that part no longer works. But it should be fixable. Either by reverting to @itsoli code, or by further refactoring.

schmunk42 commented:
I don't mind the broken tests temporarily on the develop branch.
I think it's better that to have PR like this growing and growing, so I'll go ahead and merge it...

pmk65 commented:
I updated the Playground github page with new test and examples that match the 2.x release.
